cmd/bbolt/command_surgery_cobra.go

```diff
@@ -128,3 +130,53 @@ func surgeryFreelistAbandonFunc(cmd *cobra.Command, args []string) error {
 	fmt.Fprintf(os.Stdout, "The freelist was abandoned in both meta pages.\nIt may cause some delay on next startup because bbolt needs to scan the whole db to reconstruct the free list.\n")
 	return nil
 }
+
+func newSurgeryFreelistRebuildCommand() *cobra.Command {
+	rebuildFreelistCmd := &cobra.Command{
+		Use:   "rebuild <bbolt-file> [options]",
+		Short: "Rebuild the freelist",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return errors.New("db file path not provided")
+			}
+			if len(args) > 1 {
+				return errors.New("too many arguments")
+			}
+			return nil
+		},
+		RunE: surgeryFreelistRebuildFunc,
+	}
+
+	rebuildFreelistCmd.Flags().StringVar(&surgeryTargetDBFilePath, "output", "", "path to the target db file")
+
+	return rebuildFreelistCmd
+}
+
+func surgeryFreelistRebuildFunc(cmd *cobra.Command, args []string) error {
+	srcDBPath := args[0]
+
+	// Ensure source file exists.
+	fi, err := os.Stat(srcDBPath)
+	if os.IsNotExist(err) {
+		return ErrFileNotFound
+	} else if err != nil {
+		return err
+	}
+
+	if err := common.CopyFile(srcDBPath, surgeryTargetDBFilePath); err != nil {
+		return fmt.Errorf("[freelist rebuild] copy file failed: %w", err)
+	}
+
+	// bboltDB automatically reconstruct & sync freelist in write mode.
+	db, err := bolt.Open(surgeryTargetDBFilePath, fi.Mode(), nil)
```

Comment:
To reconstruct the freelist I assume 2 conditions needs to be fullfiled: 
 1) The bolt need to be open with NoFreelistSync=false
 2) The freelist is not present in the file (i.e. was abandoned). 
I don't see a code enforcing 1) [if it's a default, I would still be explicit] nor 2).

Reply:
> 1. The bolt need to be open with NoFreelistSync=false


Makes sense & updated & thx

> 2. The freelist is not present in the file (i.e. was abandoned).

If the freelist is already present, then the command will do nothing. Please see [db.go#L273](https://github.com/etcd-io/bbolt/blob/5602d88f41e9bca30613bbf372a0f43a792afbd7/db.go#L273). Of course, doing some check beforehand will not do any harm, but I tend not to improve it which isn't necessary. WDYT?

Reply:
A.d. 2. If we call it 'rebuild' I assume we will rebuild it even if it already exists... as e.g. we think it might be wrong/corrupted after other surgeries we did. So I would rather force-trigger it by setting the freelist to 0 in the first place rather than silently skipping the request.

Reply:
> as e.g. we think it might be wrong/corrupted after other surgeries we did

If it's corrupted, the bolt.Open may fail; and we also have other command to check it.

> If we call it 'rebuild' I assume we will rebuild it even if it already exists...

I am neutral to this. Will follow your suggest to update. thx

Reply:
> If we call it 'rebuild' I assume we will rebuild it even if it already exists...

On second thought, **forcibly rebuilding** looks like a `abandon + rebuild`. Since surgery is low level command, I tend not do such enhancement. I also make the `abandon` command almost useless.  Do you still recommend forcibly rebuilding? @ptabor

Reply:
I see 2 paths: 
 1. Add a warning that the file already has freelist and that 'abandon' call is needed to force-rebuild it .
 2. Integrate it automatically. 
The 1 is closer to the principles you mentioned. I'm fine with both. Thank you.

Reply:
> Add a warning that the file already has freelist and that 'abandon' call is needed to force-rebuild it .

Sounds reasonable. Thanks @ptabor

Reply:
done. PTAL @ptabor

Reply:
thx